Refactor isReturnable method

laradev43 opened the pull request:
Order is now returnable if paid or shipped and if product return is active.

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Order is now returnable if paid or shipped and if product return is active.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Make a product return as a customer or an employee.
| UI Tests          | First commit : https://github.com/laradev43/ga.tests.ui.pr/actions/runs/12157871152 Second commit : https://github.com/laradev43/ga.tests.ui.pr/actions/runs/12254137402
| Fixed issue or discussion?     | Nothing
| Related PRs       | Nothing
| Sponsor company   | Laradev

laradev43 committed:
Refactor isReturnable method

Order is now returnable if paid or shipped and if product return is active.

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `category` should be one of these: `FO`, `BO`, `CO`, `IN`, `WS`, `TE`, `LO`, `ME` or `PM`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @laradev43!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

laradev43 commented:
The documentation says that a command must have at least one of the two cases to return a product, but in the code it didn't match. 

https://docs.prestashop-project.org/v.8-documentation/user-guide/selling/managing-customer-service/merchandise-returns#merchandisereturns-returnprocess-howthecustomerseesit

PrestaEdit reviewed (changes requested):
The documentation say this :

> The orders need to have had at least the following statuses:
At least two statuses: one with the condition "Set the order as paid" enabled and the second one with the option "Set the order as shipped" enabled.
Or
One status which has both conditions ("Set the order as paid" and "Set the order as shipped") enabled.

You need to have both condition : paid AND shipped, not only one of them.

The code is correct, there.

laradev43 commented:
Indeed, the documentation states that there should either be a status fulfilling both conditions or two statuses, each meeting one of the required conditions.

However, the original code refers to the isPaidAndShipped() method, which only checks the current status of the order. So, the current status takes precedence.

I will update my pull request to address both cases presented in the documentation.

laradev43 committed:
add configuration for product return system behavior

- Allows selection between multiple options for product returns:
  - Order delivered and paid
  - Order delivered or paid
- Provides flexibility for users to configure according to their needs.

laradev43 commented:
Hello, I've made some changes to allow product returns in three different ways.

1) The order must have been delivered and paid for.
2) The co-order must at least be delivered
3) The order must at least be paid for

This is very practical for B2B with deferred payment.
